### PR TITLE
Add replacement API for '*keys' and 'interface' key generation and import 

### LINKED ISSUE #466: Design key import/generation for the signer API
With #456 the signer API comes a lot closer to being a "fully featured API". The missing part is key generation and import (and maybe storing private keys)

Related issues: #446, #408, #280, #451 (and probably more)

## Current status (after #456):
* Using signers and keys is possible without knowing the signer or key implementation, and without using the "old" API
* **except** to create keys this sort of code is typically used:
  ```
  key = keys.generate_ed25519_key() # key is a dictionary. keys is a module we would like to start hiding from users
  pubkey = SSlibKey.from_securesystemslib_key(key) # pubkey is an implementation of Key
  signer = Signer.from_priv_key_uri("file:path/to/file?encrypted=false", pubkey) # signer is an implementation of Signer
  
  ```
* the key dictionary is only used as an intermediate step

## Proposal
* Signer implementations should provide implementation specific methods to create new keys or import them from other systems
* these methods may have whatever arguments they need
* Signer should expose the public key so callers can access it
* typical usage:
  ```
  # call implementation specific API (in this case SSlibsigner to generate a ed25519 key)
  signer = SSlibSigner.new_ed25519(...) 
  pubkey = signer.public_key
  ```
* this hides the `keys` module from users, and makes `SSlibKey.from_securesystemslib_key` unneeded
* AFAICT this should work for KMS, HSM, GPG, ...
* The implementation specific method could live elsewhere as well, but it seemed like it would make most sense in signer.

---

The part I have not 100% figured out is private key storage: how and when do you store the private key content in a way that you can then use the `from_priv_key_uri()` to load it again.  This is somewhat different depending on the key type and in many cases does not even make sense (you can't really store environment variables)... possibly it should happen during  the constructor, so `SSlibSigner.new_ed25519() ` could actually take a priv_key_uri argument and and the generated key would immediately get stored ?

### PR DESCRIPTION
**Pull request by lukpueh**
**Notes to reviewers**
 - Maybe closes #466.
 - Includes commits from #590, which can be ignored when reviewing here.
 - Is well-suited for a commit-by-commit review.
 - If desired, I can create separate smaller PRs.
 - At this point, I'd mostly like comments about whether this proposals seems fit to replace the related legacy "API".

**Background / Motivation**

My initial plan was to refactor key generation, export and import implementation in `*keys` modules, akin to signature creation (#590) and verification (#585), with the goals to
- reduce and modernize implementation code
- remove obsolete internal (schema/formats, etc.) and external (pynacl) dependencies
- preserve support for legacy on-disk key formats via `interface` (keygen, import, export) and `SSlibSigner` (sign)

Unfortunately, I couldn't find a similarly non-invasive and future-proof way, as with the signing/verifying refactors. The reason for this seems to be that on-disk key formats for public keys, and for encrypted and non-encyrypted private keys, their in-memory representation, and the public key container format and public key data format, as they are listed in tuf/in-toto metadata, are all extremely and inconsistently coupled via a loosely defined `keydict`.

Re-writing large portions of legacy code, just to support inconsistent, proprietary on-disk key formats, didn't seem worth the effort. Instead, I added a non-backwards-compatible but modern and light-weight replacement API, which supports [old and new](https://theupdateframework.github.io/python-tuf/2023/01/24/securesystemslib-signer-api.html) securesystemslib usage paradigms for the known key types.

IMPORT NOTE: Public key metadata formats as used in TUF and in-toto metadata are not changed and remain backwards compatible.

**Description of changes**

Add methods for ketypes rsa, ecdsa and ed25519 to:

* generate new key pair as `Signer` instance
  *for simple in-memory setups, tests or demos*

* import public key from standard PEM/subjectPublicKeyInfo encoded file as `Key` instance
  *for simple setups, tests and demos, which use stand-alone key files*

* import encrypted or non-encrypted private key from standard PEM/PKCS8 encoded file as `Signer` instance
  *for simple setups, tests and demos, which use stand-alone key file;  but also compatible with new usage-paradigm, with signer uri*

Missing features compared to legacy API
* Generate and write key pairs
* Import key files from proprietary legacy formats
* Import private key without already having the public key 
* Import signer from envvar

**Future work / TODO** (not necessarily in this PR)

* Polish implementation: I think, the API design is right and complete, but internally we could do some deduplication and polishing (e.g. keytype/scheme switches)
* Use constants instead of hard-coded keytype/scheme strings and lists (#593)
* Review interface boundaries between CryptoSigner and SSlibKey (e.g. CrytpoSigner instances require an SSlibKey, but not all SSlibKey schemes are supported by all signers)
* Review validation, e.g. do we fail on
    - incorrect scheme/type for signer?
    - incorrect private key for signer?
    - incorrect public key values for scheme/type?
    - missing optional dependency (UnsupportedLibraryError) in all public methods?
* More tests

### CODE REVIEW

**Comment by lukpueh** [comment]
Some additional comments wrt:

> Missing features compared to legacy API
> 
> * Generate and write key pairs

Now that file-import works with standard formats, users can easily use 3rd-party tools for this. I suggest to just add e.g. the `openssl` commands, which I used to generate test key files (see related commit message), to the docs.

If there's a need, we can easily add the feature. It's just a few lines of pyca/cryptography code).

> * Import key files from proprietary legacy formats

Let's just not support this in new API. Instead, I suggest to offer migration support when releasing a version without the legacy interfaces.

> * Import private key without already having the public key
> * Import signer from envvar

Same as for generate and write, if there's a need for these features we can easily add them.

**Review by jku — COMMENTED** [review]
I think this is close... Left a few comments while reading, but the core issue I have is that the Signer class hierarchy is now quite complicated, the hierarchy is largely public,  and it's unclear (to me) what classes the caller should be using in different situations. In other words: why is this class public, what is its purpose for the user. This could just boil down to me not understanding the use cases these are for -- each public class could document when it should be instantiated and how.

I also understand that not everything is implemented... but fixing that is hard if it's not clear what needs to be implemented

I guess the use cases where we should be able to point to a specific class and method are
1. How do I construct a signer when I need to sign? (I assume the answer is still `Signer.from_private_key_uri()` should be used, and SIGNER_FROM_URI_SCHEME can be used to configure the implementations used in the application)
2. How do I construct a private key uri and Key for storing for later use when using cryptography (assume we have bytes we need from cryptography) -- equivalent of `import_()` in other Signers. A complete example would help
3. generating a new private and public key for storing a private key uri, private key and public key (I'd really like to leave this to user calling cryptography and giving us bytes, but... I suppose we need to support it. I'm less sure if the classes need to be public or ig these are just public helper methods -- does showing the class to user give some value?)

TL;DR: I think it's close but I still don't quite get it, we should probably chat and maybe I'll figure it out...

File: securesystemslib/signer/_signer.py [hunk 1]
```diff
@@ -225,5 +252,294 @@ def sign(self, payload: bytes) -> Signature:
                 securesystemslib.exceptions.UnsupportedAlgorithmError:
                 Signing errors.
         """
-        sig_dict = sslib_keys.create_signature(self.key_dict, payload)
-        return Signature(**sig_dict)
+        return self._crypto_signer.sign(payload)
+
+
+class CryptoSigner(Signer, metaclass=ABCMeta):
```
[the remark this answers is not in the document]
Comment (lukpueh) [reply securesystemslib/signer/_signer.py:258]: > Is SSlibSigner just a legacy component now?

Correct. I thought I mentioned this somewhere, but maybe I didn't. It looks like I only complained about the inconsistent and proprietary on-disk and in-memory data structure, which is the meat of SSlibSigner. I just couldn't find a way to fix it, without breaking the SSlibSigner interface.

So, yes, I do consider CryptoSigner to be a modern replacement for SSlibSigner use cases. And for a transition period, CryptoSigner can also provide the implementation for SSlibSigner, which would allow us to deprecate `key*` independently of SSlibSigner. I'm okay to revert the latter part, if that only makes things more complicated.

Regarding class hierarchy of CryptoSigner, I agree that the subclasses could be non-public API. The only reason why I made them public, is to namespace the corresponding generate methods: `RSASigner.generate`, etc. But we could also expose these directly on CryptoSigner, e.g. `CryptoSigner.generate_rsa`, etc.

[the remark this answers is not in the document]
Comment (lukpueh) [reply securesystemslib/signer/_signer.py:258]: > > Is SSlibSigner just a legacy component now?
> 
> Correct. I thought I mentioned this somewhere, but maybe I didn't. 

Mentioned it in: https://github.com/secure-systems-lab/securesystemslib/pull/590#issuecomment-1638176113. Should have mentioned it in this PR too.

File: securesystemslib/signer/__init__.py [hunk 2]
```diff
@@ -15,14 +16,14 @@
...
 from securesystemslib.signer._sigstore_signer import SigstoreKey, SigstoreSigner
 from securesystemslib.signer._spx_signer import (
     SpxKey,
     SpxSigner,
     generate_spx_key_pair,
 )
+from securesystemslib.signer._sslib_signer import SSlibSigner
 
 # Register supported private key uri schemes and the Signers implementing them
 SIGNER_FOR_URI_SCHEME.update(
```
Comment (jku) [inline securesystemslib/signer/__init__.py:29]: Should we change the default handler for FILE_URI_SCHEME? (and maybe leave a comment that ENVVAR_URI_SCHEME is deprecated unless someone reimplements it)

**Review by jku — APPROVED** [review]
Yeah this looks good to me. Would love to have some other peoples opinions but I'm happy with this.

Left a question on updating the default signer lookup table
